py/vtdb/grpc_vtgate_client.py

```diff
@@ -0,0 +1,536 @@
+"""
+
+import datetime
+import logging
+import re
+from urlparse import urlparse
+
+from grpc.beta import implementations
+from grpc.beta import interfaces
+from grpc.framework.interfaces.face import face
+
+from vtproto import query_pb2
+from vtproto import topodata_pb2
+from vtproto import vtgate_pb2
+from vtproto import vtgateservice_pb2
+
+from vtdb import dbapi
+from vtdb import dbexceptions
+from vtdb import field_types
+from vtdb import field_types_proto3
+from vtdb import keyrange_constants
+from vtdb import keyspace
+from vtdb import times
+from vtdb import vtdb_logger
+from vtdb import vtgate_client
+from vtdb import vtgate_cursor
+from vtdb import vtgate_utils
+
+
+INT_UPPERBOUND_PLUS_ONE = 1<<63
+_errno_pattern = re.compile(r'\(errno (\d+)\)', re.IGNORECASE)
+_throttler_err_pattern = re.compile(
+    r'exceeded (.*) quota, rate limiting', re.IGNORECASE)
+
+
+class GRPCVTGateConnection(vtgate_client.VTGateClient):
+  """A simple, direct connection to the vtgate query service.
+  """
+
+  def __init__(self, addr, timeout):
+    self.addr = addr
+    self.timeout = timeout
+    self.stub = None
+    self.session = None
+    self.logger_object = vtdb_logger.get_logger()
+
+  def dial(self):
+    if self.stub:
+      self.stub.close()
+
+    p = urlparse('http://' + self.addr)
+    channel = implementations.insecure_channel(p.hostname, p.port)
+    self.stub = vtgateservice_pb2.beta_create_Vitess_stub(channel)
+
+  def close(self):
+    self.stub = None
+
+  def is_closed(self):
+    return self.stub is None
+
+  def cursor(self, *pargs, **kwargs):
+    cursorclass = kwargs.pop('cursorclass', None) or vtgate_cursor.VTGateCursor
+    return cursorclass(self, *pargs, **kwargs)
+
+  def begin(self, effective_caller_id=None):
+    try:
+      request = vtgate_pb2.BeginRequest()
+      _add_caller_id(request, effective_caller_id)
+      response = self.stub.Begin(request, self.timeout)
+      self.effective_caller_id = effective_caller_id
+      self.session = response.session
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      raise _convert_exception(e)
+
+  def commit(self):
+    try:
+      request = vtgate_pb2.CommitRequest()
+      _add_caller_id(request, self.effective_caller_id)
+      self._add_session(request)
+      self.stub.Commit(request, self.timeout)
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      raise _convert_exception(e)
+    finally:
+      self.session = None
+      self.effective_caller_id = None
+
+  def rollback(self):
+    try:
+      request = vtgate_pb2.RollbackRequest()
+      _add_caller_id(request, self.effective_caller_id)
+      self._add_session(request)
+      self.stub.Rollback(request, self.timeout)
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      raise _convert_exception(e)
+    finally:
+      self.session = None
+      self.effective_caller_id = None
+
+  def _execute(self, sql, bind_variables, keyspace_name, tablet_type,
```

Comment:
This should have a more detailed docstring, maybe just copy the one from the GoRPC client?